rust/chains/hyperlane-cosmos/src/providers/rpc.rs

```diff
@@ -84,84 +87,116 @@ impl WasmIndexer for CosmosWasmIndexer {
-                    )
-                    .unwrap()
-                    .as_bytes(),
-                )
-            })
-            .collect();
-
-        let mut result: Vec<(T, LogMeta)> = vec![];
-        let tx_results = block_result.txs_results;
-
-        if tx_results.is_none() {
-            return Ok(result);
+        let contract_address = self.get_contract_addr()?;
+
+        let block_step: u32 = 100;
+        let mut block_hash_vec: Vec<(u64, H256)> = vec![];
+        let block_query: Query = format!(
+            "block.height >= {} AND block.height <= {}",
+            range.start(),
+            range.end(),
+        )
+        .parse()
+        .unwrap();
+        let total_block_count = range.end() - range.start() + 1;
+        let last_block_page = total_block_count / 100 + (total_block_count % 100 != 0) as u32;
+
+        for _ in 1..=last_block_page {
+            let blocks = client
+                .block_search(block_query.clone(), 1, block_step as u8, Order::Ascending)
+                .await?;
+
+            block_hash_vec.extend(
+                blocks
+                    .blocks
+                    .iter()
+                    .map(|b| {
+                        (
+                            b.block.header.height.value(),
+                            H256::from_slice(b.block_id.hash.as_bytes()),
+                        )
+                    })
+                    .collect::<Vec<(u64, H256)>>(),
+            );
         }
-
-        let addr = self.get_contract_addr()?;
-
-        for (idx, tx) in tx_results.unwrap().iter().enumerate() {
-            let tx_hash = tx_hash[idx];
-            if tx.code.is_err() {
-                debug!("tx {:?} has failed. skip!", tx_hash);
-                continue;
-            }
-
-            let mut available = false;
-
-            let mut parse_result: Vec<(T, LogMeta)> = vec![];
-
-            let logs = serde_json::from_str::<Vec<Events>>(&tx.log)?;
-            let logs = logs.first().unwrap();
-
-            for (log_idx, event) in logs.events.clone().into_iter().enumerate() {
-                if event.typ.as_str().starts_with(Self::WASM_TYPE)
-                    && event.attributes[0].value == addr
-                {
-                    available = true;
-                } else if event.typ.as_str() != self.event_type {
+        let block_hash: HashMap<u64, H256> = block_hash_vec.into_iter().collect();
+
+        // Page starts from 1
+        let query = Query::default()
+            .and_gte("tx.height", *range.start() as u64)
+            .and_lte("tx.height", *range.end() as u64)
+            .and_eq(
+                format!("{}-{}._contract_address", Self::WASM_TYPE, self.event_type),
+                contract_address.clone(),
+            );
+
+        println!("Query: {:?}", query.to_string());
+
+        let tx_search_result = client
+            .tx_search(query.clone(), false, 1, 30, Order::Ascending)
```

Comment:
Similarly here, I would prefer removing the pagination logic if possible

Reply:
Looks like the max returned per RPC is 100 https://docs.tendermint.com/v0.34/rpc/#/Info/tx_search so having pagination logic feels reasonable

Any reason to be using a `per_page` of 30 instead of 100? May also be good to put it into a const so that we don't have a few different places with 30 hardcoded

Reply:
@daniel-savu @tkporter yup there's max returned per RPC is 100 so it necessary for paging logic. I restrict the per_page as 30 because of the transfer payload size can be too big to cause other unexpected errors.